lib/semantics/check-select-stmt.cc

```diff
@@ -0,0 +1,239 @@
+//===-- lib/semantics/check-select-stmt.cc --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "tools.h"
+#include "../parser/message.h"
+
+namespace Fortran::semantics {
+
+void SelectStmtChecker::Leave(
+    const parser::CaseConstruct &selectCaseConstruct) {
+  const auto &selectCaseStmt{
+      std::get<parser::Statement<parser::SelectCaseStmt>>(
+          selectCaseConstruct.t)};
+  const auto &parsedExpr{
+      std::get<parser::Scalar<parser::Expr>>(selectCaseStmt.statement.t).thing};
+  std::optional<TypeCategory> selectCaseStmtType;
+  bool validCaseStmtType{false};
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (auto type{expr->GetType()}) {
+      selectCaseStmtType = type->category();
+    }
+  }
+  if (!selectCaseStmtType ||
+      ((selectCaseStmtType.value() != TypeCategory::Integer) &&
+          (selectCaseStmtType.value() != TypeCategory::Character) &&
+          (selectCaseStmtType.value() != TypeCategory::Logical))) {  // C1145
+    context_.Say(parsedExpr.source,
+        "SELECT CASE expression must be of type CHARACTER, INTERGER, OR LOGICAL"_err_en_US);
+  } else {
+    validCaseStmtType = true;
+  }
+
+  const auto &caseList{
+      std::get<std::list<parser::CaseConstruct::Case>>(selectCaseConstruct.t)};
+  bool defaultCaseFound{false};
+  using var_t = std::variant<std::int64_t, bool, std::string>;
+  using IntType = std::vector<
+      std::pair<std::optional<std::int64_t>, std::optional<std::int64_t>>>;
+  using StringType = std::vector<
+      std::pair<std::optional<std::string>, std::optional<std::string>>>;
+  std::set<var_t> caseVals;
+  std::pair<std::set<var_t>::iterator, bool> ret;
+  IntType caseRangeIntVals;
+  IntType::iterator intIter;
+  StringType caseRangeStringVals;
+  StringType::iterator stringIter;
+
+  for (const auto &cases : caseList) {
+    const auto &caseStmt{
+        std::get<parser::Statement<parser::CaseStmt>>(cases.t)};
+    const auto &caseSelector{
+        std::get<parser::CaseSelector>(caseStmt.statement.t)};
+    if (std::holds_alternative<parser::Default>(caseSelector.u)) {
+      if (!defaultCaseFound) {
+        defaultCaseFound = true;
+      } else {  // C1146 (R1140)
+        context_.Say(caseStmt.source,
+            "Not more than one of the selectors of case statements may be default"_err_en_US);
```

Comment:
I  suggest replacing "case statements" with "SELECT CASE" or "SELECT CASE statement" like in other errors.